### PR TITLE
docs: Add AWS Resource Naming section

### PR DESCRIPTION
**Pull request by dipasqualew**
As discussed on the Engineering Forum (19/11/2020) this is a naming
convention we could use for all new AWS resources.

See:
- comicrelief/engineering-forum#3

### CODE REVIEW

**Review by adamclark-dev — APPROVED** [review]

File: backend/coding-standards.md [hunk 1]
```diff
@@ -24,6 +24,32 @@ For our PHP projects we use [PHPCS](https://github.com/squizlabs/PHP_CodeSniffer
 
 We also use [PHPCD](https://github.com/sebastianbergmann/phpcpd) to ensure that code is not duplicated within the project.
 
+## AWS Resource Naming
+
+Historic resources don't follow an agreed convention. However, for all new resources, both in old and new projects, we should follow the schema:
+
+`${service-name}-${service-stage}-${resource-type}-${resource-label}`
```
Comment (tomisaacmarshall) [inline backend/coding-standards.md:31]: Is it worth using underscores as separators? Is this allowed for AWS resources? Just to differentiate between dashes in names

Also curious to know why it's useful to have the resource type in the name I suppose. Do all S3 buckets really need to contain S3 in the name?

Also would be tempted to change it to 

`${service-name}-${resource-label}-${service-stage}`

or

`${service-name}-${resource-type}-${resource-label}-${service-stage}`

Basically putting the label before the stage, so that the staging/production versions of a given resource are grouped together when sorting by name, rather than grouping things by stage

Comment (dipasqualew) [reply backend/coding-standards.md:31]: There are a few resources that don't support underscores.

What do everyone else think about having `stage` after `label`? To me there are two point to consider

1. We define a `prefix` in SLS so we can actual name things by using `${self:custom.prefix}` - which makes things easier to manage and to maintain
2. I really like having all staging and all production grouped together, so there can be no mix.

Comment (tomisaacmarshall) [reply backend/coding-standards.md:31]: Just a matter of taste I guess so I withdraw the suggestion. I more often want to look at the two versions of a given s3 bucket for example, rather than see all my staging buckets together - but there's always filtering!

Comment (lebaz20) [inline backend/coding-standards.md:31]: I have only one concern, as AWS has limits on name length for pretty much everything, We may not be able to apply this rule everywhere.
I would opt for the shortest significant naming, so `${service-name}-${service-stage}` as a starter, then ${resource-label} would be added if needed.

Comment (dipasqualew) [reply backend/coding-standards.md:31]: I disagree! I would advise to go towards the opposite route, with explicit, long names by default and only truncating where the resource goes over the maximum allowed by AWS (which is 80, isn't it?)

Even in lengthy service names we don't get close to 80, for example:
- `serverless-prize-platform-production-dynamodb-prize-categories` is 62 characters

Comment (dipasqualew) [reply backend/coding-standards.md:31]: However, if we want to save characters, I would advise to change the `serverless` prefix into `sls`, which is a considerable saving

Comment (lebaz20) [reply backend/coding-standards.md:31]: Ok, won't argue with either route to take.
It's an extreme case, where the limitation will dictate the naming anyway, and unfortunately it's sometimes less than 80, 64 in some instances!
I would just prefer to avoid redundant info in name, like dynamodb name including `dynamodb`!

Comment (dipasqualew) [reply backend/coding-standards.md:31]: I think we can skip the resource type in the name, it was there just to highlight what the resource is by just reading the name. The use case where it makes sense is correlated resources, such as

- `serverless-prize-platform-staging-s3-tickets`
- `serverless-prize-platform-staging-glue-tickets`

Or

- `supporter-event-service-staging-sqs-main`
- `supporter-event-service-staging-sqs-dlq-main`

I am not precious about the resource type, but I think the label is important, even if it is just `-main` (+5 characters)

Comment (tomisaacmarshall) [reply backend/coding-standards.md:31]: Was also wondering if char limits could be a concern, or even just UIs where it's hard to read long names.

However, `resource-label` seems like the most important part of the name to me and don't think it can be dropped. It's the key identifier isn't it? `sls-contact-store-staging-s3` is pretty useless as a name.

Still don't get the need for the resource type (s3, dynamodb) as I'm not aware of situations when we'd be scanning the names of resources without knowing what type we're looking at, although that's not to say there aren't any!

Comment (tomisaacmarshall) [reply backend/coding-standards.md:31]: Actually I take that back. I guess there probably will be lots of times we only want _one_ 'resource' e.g. S3 bucket for a given service.... :)

Comment (dipasqualew) [reply backend/coding-standards.md:31]: It is also because I had the case where I named a S3 bucket and a Glue Table the same way and in the failed deploys I kept working on the table, totally forgot about the bucket, spent a lot of time with an unhelpful error from serverless (my fault for misleading naming).

See my example above. How would we differentiate a s3 bucket and the related glue table? They can't share the same name.

`service`, `stage` and `label` are the fundamental bits. If the resulting string is too long, the `service` bit is probably the bit we need to make shorter (for future services)

Comment (seb-cr) [reply backend/coding-standards.md:31]: > However, if we want to save characters, I would advise to change the serverless prefix into sls, which is a considerable saving

We could get rid of `serverless-` completely, since every service we deploy will be serverless... I think the GitHub repo name should still be `serverless-prize-platform`, but the service name as defined in `serverless.yml` could omit it.

Pro: saves 11 characters; avoids noise in names; some AWS filtering works on prefix only, so currently always have to always type "serverless".
Con: service name doesn't exactly match repo name, but that would happen anyway if we opted for `sls`, and GitHub's search is not prefix-based.
